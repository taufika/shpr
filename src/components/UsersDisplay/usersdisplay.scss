.users-display-main {
  width: calc(100vw - 255px)
}

.ready {
  padding: 32px;

  .cards-place {
    display: flex;
    margin-top: 32px;
    width: 100%;
    overflow-x: auto;
    padding-bottom: 40px;
  }

  .pagin-place {
    margin-top: 20px;
    margin-bottom: 20px;
    display: flex;
    justify-content: center;

    button.page {
      width: 160px;
      border: 0;
      background: none;
      cursor: pointer;

      .fas {
        margin-left: 12px;
        margin-right: 12px;
      }

      &:disabled {
        cursor: not-allowed;
      }
    }
  }
}


.loading, .error {
  position: relative;
  height: 100%;

  span {
    color: rgb(80, 80, 80);
    display: inline-block;
    min-width: 120px;
    position: absolute;
    left: 50%;
    top: 20%;
    transform: translateX(-50%);
  }
}