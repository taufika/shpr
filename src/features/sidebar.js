import { createSlice } from '@reduxjs/toolkit'

export const sidebarSlice = createSlice({
  name: 'users',
  initialState: {
    isSidebarActive: false,
  },
  reducers: {
    toggleSidebar: (state) => {
      state.isSidebarActive = !state.isSidebarActive;
    },
  }
});

export const { toggleSidebar } = sidebarSlice.actions;

export default sidebarSlice.reducer;