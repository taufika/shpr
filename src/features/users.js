import { createSlice } from '@reduxjs/toolkit'

export const usersSlice = createSlice({
  name: 'users',
  initialState: {
    users: [],
    status: 'ready',
  },
  reducers: {
    setLoading: (state) => {
      state.status = 'loading';
    },
    setReady: (state) => {
      state.status = 'ready';
    },
    setError: (state) => {
      state.status = 'error';
    },
    assignUsers: (state, action) => {
      state.users = action.payload;
    },
  }
});

export const { assignUsers, setError, setReady, setLoading } = usersSlice.actions;

export default usersSlice.reducer;