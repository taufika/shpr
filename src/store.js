import { configureStore } from '@reduxjs/toolkit'
import usersReducer from './features/users';
import sidebarReducer from './features/sidebar';

export default configureStore({
  reducer: {
    users: usersReducer,
    sidebar: sidebarReducer,
  },
})
