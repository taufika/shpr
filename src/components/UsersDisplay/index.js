import React from 'react';
import { connect } from 'react-redux'
import { useSearchParams } from 'react-router-dom';

import SearchBar from '../SearchBar';
import UserCard from '../UserCard';

import './usersdisplay.scss';

class UsersDisplay extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      offset: parseInt(this.props.router.searchParams.get('offset') || 0, 10),
      keyword: this.props.router.searchParams.get('keyword') || '',
    }
    this.handleKeywordChange = this.handleKeywordChange.bind(this);
    this.getFilteredUsers = this.getFilteredUsers.bind(this);
    this.getUsersOnOffset = this.getUsersOnOffset.bind(this);
    this.handleChangePage = this.handleChangePage.bind(this);
  }

  handleKeywordChange(keyword) {
    this.setState({ keyword, offset: 0 });
    if (keyword) {
      this.props.router.setSearchParams({ keyword });
    } else {
      this.props.router.setSearchParams({ });
    }
  }

  getFilteredUsers() {
    return this.props.users.filter(u => u.name.first.toLowerCase().includes(this.state.keyword.toLowerCase()));
  }

  getUsersOnOffset(offset, size) {
    return this.getFilteredUsers().slice(offset, offset + size);
  }

  handleChangePage(newOffset) {
    this.setState({ offset: newOffset });
    this.props.router.setSearchParams({ keyword: this.state.keyword, offset: newOffset });
  }

  render() {
    let section;
    const pageSize = 5;

    switch (this.props.status) {
      case 'ready':
        section = <div className='ready'>
          <SearchBar
            keyword={this.state.keyword}
            keywordHandler={this.handleKeywordChange}
          />
          <div className='cards-place'>
            {
              this.getUsersOnOffset(this.state.offset, pageSize).map((u,idx) => 
                <UserCard key={idx + u.name.first} user={u} idx={idx} />)
            }
          </div>
          <div className='pagin-place'>
            <button
              data-testid='prev-button'
              className='prev page'
              disabled={this.state.offset === 0} 
              onClick={() => this.handleChangePage(this.state.offset - pageSize)}>
              <i className="fas fa-chevron-left" />
              <span>Previous Page</span>
            </button>
            <button
              data-testid='next-button'
              className='next page'
              disabled={this.state.offset + pageSize >= this.getFilteredUsers().length }
              onClick={() => this.handleChangePage(this.state.offset + pageSize)}>
              <span>Next Page</span>
              <i className="fas fa-chevron-right" />
            </button>
          </div>
        </div>;
        break;
      case 'error':
        section = <div className='error'>
          <span>Gagal memuat data. Silakan muat ulang halaman.</span>
        </div>
        break;
      default:
        section = <div className='loading'>
          <span>Memuat Data...</span>
        </div>
    }
    return <div className='users-display-main'>
      {section}
    </div>;
  };
}

function UserDisplayWrapper(Component) {
  function UserDisplayWithRoute(props) {
    let [searchParams, setSearchParams] = useSearchParams();
    return (
      <Component 
        {...props}
        router={{ searchParams, setSearchParams }}
      />
    );
  }

  return UserDisplayWithRoute;
}

const propsMatcher = state => ({ users: state.users.users, status: state.users.status });
export default UserDisplayWrapper(connect(propsMatcher)(UsersDisplay));
