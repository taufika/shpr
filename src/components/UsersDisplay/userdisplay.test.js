import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { useEffect } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { assignUsers, setError, setLoading, setReady } from '../../features/users'
import { useSearchParams } from 'react-router-dom';

import store from '../../store';

import UsersDisplay from '.';


const wrapped = () => (
  <Provider store={store}>
    <BrowserRouter>
      <UsersDisplay />
    </BrowserRouter>
  </Provider>
);

const dummyUser = {
  name: {
    first: 'Taufik',
    last: 'Akbar',
  },
  cell: '082129136223',
  email: 'taufik.akbar7@gmail.com',
  dob: {
    date: '1994-05-17',
  },
  picture: {
    medium: 'https://some.picture.url',
  },
};

describe('Component:UserDisplay', () => {
  it('Should render and match snapshot: empty users', () => {
    const { container } = render(wrapped());
    expect(container).toMatchSnapshot();
  });

  it('Should render and match snapshot: loading', () => {
    const { container } = render(wrapped());
    store.dispatch(setLoading());
    expect(container).toMatchSnapshot();
  });

  it('Should render and match snapshot: load error', () => {
    const { container } = render(wrapped());
    store.dispatch(setError());
    expect(container).toMatchSnapshot();
  });

  it('Should render properly: 4 users', () => {
    render(wrapped());
    store.dispatch(assignUsers(
      Array(4)
        .fill(dummyUser)
    ));
    store.dispatch(setReady());
    const cards = screen.getAllByTestId('user-card');
    expect(cards.length).toBe(4);
    expect(screen.getByTestId('prev-button').disabled).toBe(true);
    expect(screen.getByTestId('next-button').disabled).toBe(true);
  });

  it('Should render properly: 10 users', () => {
    render(wrapped());
    store.dispatch(assignUsers(
      Array(10)
        .fill(dummyUser)
    ));
    store.dispatch(setReady());
    const cards = screen.getAllByTestId('user-card');
    expect(cards.length).toBe(5);
    expect(screen.getByTestId('prev-button').disabled).toBe(true);
    expect(screen.getByTestId('next-button').disabled).toBe(false);
  });

  it('Should let user go next page and set offset on url search query', () => {
    render(wrapped());
    store.dispatch(assignUsers(
      Array(7)
        .fill(dummyUser)
    ));
    store.dispatch(setReady());
    const prevPage = screen.getByTestId('prev-button');
    const nextPage = screen.getByTestId('next-button');
    fireEvent.click(nextPage);

    // 7 items so show 2 on page 2
    const cards = screen.getAllByTestId('user-card');
    expect(cards.length).toBe(2);
    expect(prevPage.disabled).toBe(false);
    expect(nextPage.disabled).toBe(true);
    expect(window.location.search).toContain('offset=5');
  });

  it('Should set keyword url search query on search', () => {
    render(wrapped());
    store.dispatch(assignUsers(
      Array(7)
        .fill(dummyUser)
    ));
    store.dispatch(setReady());
    const input = screen.getByTestId('search-input');
    fireEvent.change(input, { target: { value: 'search keyword' } });
    expect(window.location.search).toContain('keyword=search+keyword');
  });

  it('Should let user delete keyword and then remove query from url', () => {
    store.dispatch(assignUsers(
      Array(7)
        .fill(dummyUser)
    ));
    store.dispatch(setReady());
    render(wrapped());
    const input = screen.getByTestId('search-input');
    fireEvent.change(input, { target: { value: 'test' } });
    expect(window.location.search).toContain('keyword=test');
    fireEvent.change(input, { target: { value: '' } });
    expect(window.location.search).not.toContain('keyword');
  });

  it('Should let user go next page and back to previous page', () => {
    render(wrapped());
    store.dispatch(assignUsers(
      Array(7)
        .fill(dummyUser)
    ));
    store.dispatch(setReady());
    const prevPage = screen.getByTestId('prev-button');
    const nextPage = screen.getByTestId('next-button');
    // go to page 2
    fireEvent.click(nextPage);
    // go back to page 1
    fireEvent.click(prevPage);

    const cards = screen.getAllByTestId('user-card');
    expect(cards.length).toBe(5);
    expect(prevPage.disabled).toBe(true);
    expect(nextPage.disabled).toBe(false);
    expect(window.location.search).toContain('offset=0');
  });

  it('Should get component state from url search query on load', () => {
    Object.defineProperty(window, 'location', {
      value: {
        search: '?keyword=taufik&offset=5',
      },
    });
    store.dispatch(assignUsers(
      Array(7)
        .fill(dummyUser)
    ));
    store.dispatch(setReady());
    render(wrapped());
    const input = screen.getByTestId('search-input');
    const prevPage = screen.getByTestId('prev-button');
    const nextPage = screen.getByTestId('next-button');
    expect(input.value).toBe('taufik');
    expect(prevPage.disabled).toBe(false);
    expect(nextPage.disabled).toBe(true);
  });
});
