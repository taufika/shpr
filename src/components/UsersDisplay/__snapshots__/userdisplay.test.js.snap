// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:UserDisplay Should render and match snapshot: empty users 1`] = `
<div>
  <div
    class="users-display-main"
  >
    <div
      class="ready"
    >
      <div
        class="search-bar"
      >
        <div
          class="title-area"
        >
          <h1>
            Driver Management
          </h1>
          <h3>
            Data driver yang bekerja dengan Anda.
          </h3>
        </div>
        <div
          class="search-area"
        >
          <div
            class="search-icon"
          >
            <i
              class="fas fa-search"
            />
          </div>
          <input
            data-testid="search-input"
            placeholder="Cari Driver"
            value=""
          />
          <button
            class="add-driver"
          >
            Tambah Driver
            <i
              class="fas fa-plus"
            />
          </button>
        </div>
      </div>
      <div
        class="cards-place"
      />
      <div
        class="pagin-place"
      >
        <button
          class="prev page"
          data-testid="prev-button"
          disabled=""
        >
          <i
            class="fas fa-chevron-left"
          />
          <span>
            Previous Page
          </span>
        </button>
        <button
          class="next page"
          data-testid="next-button"
          disabled=""
        >
          <span>
            Next Page
          </span>
          <i
            class="fas fa-chevron-right"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Component:UserDisplay Should render and match snapshot: load error 1`] = `
<div>
  <div
    class="users-display-main"
  >
    <div
      class="error"
    >
      <span>
        Gagal memuat data. Silakan muat ulang halaman.
      </span>
    </div>
  </div>
</div>
`;

exports[`Component:UserDisplay Should render and match snapshot: loading 1`] = `
<div>
  <div
    class="users-display-main"
  >
    <div
      class="loading"
    >
      <span>
        Memuat Data...
      </span>
    </div>
  </div>
</div>
`;
